File: ios/TwilioStereoTonePlayer.m
//
//  RCTTWAudioPlayer.m
//  RNTwilioVideoWebRTC
//
//  Created by Umar Nizamani on 29/06/2020.
//

#import "TwilioStereoTonePlayer.h"
 
@implementation TwilioStereoTonePlayer

-(id)init {
    _volume = 1.0f;
    _playbackSpeed = 1.0f;
    return self;
}

static RCTTWCustomAudioDevice *_audioDevice;

+ ( RCTTWCustomAudioDevice *)audioDevice {
   if (_audioDevice==nil) {
      [TwilioStereoTonePlayer setAudioDevice:_audioDevice];
   }
   return _audioDevice;
}

+ (void)setAudioDevice:(RCTTWCustomAudioDevice *)audioDevice {
    _audioDevice = audioDevice;
}


RCT_EXPORT_MODULE()

RCT_EXPORT_METHOD(initialize:(int)maxLoadableFiles) {
    _maxLoadableFiles = maxLoadableFiles;
    _loadedFiles = [[NSMutableDictionary alloc]initWithCapacity:maxLoadableFiles];
}

RCT_EXPORT_METHOD(preload:(NSString*)filename resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    NSLog(@"preload %@", filename);
    
    NSError *err;
    
    // Sanity Check
    if (_loadedFiles == NULL) {
        reject(@"Error", [NSString stringWithFormat:@"Please call initialize first"], err);
        return;
    }
    
    if ([_loadedFiles count] >= _maxLoadableFiles) {
        reject(@"Error", [NSString stringWithFormat:@"Trying to load more files than you called initialize with! Already loaded %i files", _maxLoadableFiles], err);
        return;
    }
    
    // If we have already loaded this file then just instantly resolve the promise to be true
    if ([_loadedFiles objectForKey:filename] != NULL) {
        resolve(@(true));
        return;
    }

    // Create a file path based on the local bundle as we only expect to play files that are locally on disk
    NSString *filepath = [NSString stringWithFormat:@"%@/%@", [[NSBundle mainBundle] bundlePath], filename];
    
    // Make sure the path is escaped correctly and only has strings that are supported in a file path
    NSCharacterSet *set = [NSCharacterSet URLFragmentAllowedCharacterSet];
    NSString * filepathEscaped = [filepath stringByAddingPercentEncodingWithAllowedCharacters:set];

    // Create a local file URL
    NSURL *url = [NSURL fileURLWithPath:filepath];
    
    // Try to load the AVAudio File
    AVAudioFile *file = [[AVAudioFile alloc] initForReading:url error:&err];
    if (err != NULL) {
        reject(@"Error", [NSString stringWithFormat:@"Failed to load file at path: %@", filepathEscaped], err);
        return;
    }
    
    // Create an audio buffer to read the file into
    AVAudioPCMBuffer *musicBuffer = [[AVAudioPCMBuffer alloc] initWithPCMFormat:file.processingFormat
                                                 frameCapacity:(AVAudioFrameCount)file.length];

    // Lets try to load the file into an audio buffer
    BOOL success = [file readIntoBuffer:musicBuffer error:&err];
    if (!success) {
        reject(@"Error", [NSString stringWithFormat:@"Failed to read file into audio buffer at path: %@", filepathEscaped], err);
        return;
    }
    
    [_loadedFiles setObject:musicBuffer forKey:filename];

    resolve(@(true));
}

RCT_EXPORT_METHOD(play:(NSString*)filename isLooping:(BOOL)isLooping volume:(float)volume playbackSpeed:(float)playbackSpeed resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    NSLog(@"play %@ - volume %f - isLooping %s", filename, volume, isLooping ? "Yes" : "No");
    
    NSError *err;
    
    // Sanity Check
    if (_loadedFiles == NULL) {
        reject(@"Error", [NSString stringWithFormat:@"Please call initialize first"], err);
        return;
    }
    
    if (TwilioStereoTonePlayer.audioDevice == nil) {
        reject(@"Error", [NSString stringWithFormat:@"No Custom Audio device found! Are you trying to call play before the Twilio Local Audio Track is created or running in no custom device mode?"], err);
        return;
    }
    
    // Try to load the preloaded player from the loaded players list
    AVAudioPCMBuffer* fileToPlay = [_loadedFiles objectForKey:filename];
    
    // If the current file is not already loaded
    if (fileToPlay == nil) {
        
        // Then make sure we load the file first
        [self preload:filename resolver:resolve rejecter:reject];
        
        fileToPlay = [_loadedFiles objectForKey:filename];
        
        // If we still can't find the file, then there was probably an error loading the file.
        // We will instantly return as we expect preload to already have called the rejection promise.
        if (fileToPlay == nil) {
            return;
        }
    }

    // Before we play a new tone, lets make sure we pause any tone that is currently playing
    [self pause];
    
    // Try to schedule playback on the Twilio Custom Audio Device
    [[TwilioStereoTonePlayer audioDevice] playBuffer:fileToPlay isLooping:isLooping volume:volume playbackSpeed:playbackSpeed];
    
    _volume = volume;
    _playbackSpeed = playbackSpeed;
    _currentPlayingFile = filename;
    
    resolve(@(true));
}

RCT_EXPORT_METHOD(pause) {
    NSLog(@"pause");
    
    if (TwilioStereoTonePlayer.audioDevice == nil) {
        NSLog(@"Error: No Custom Audio device found! Are you trying to call pause before the Twilio Local Audio Track is created or running in no custom device mode?");
        return;
    }
    
    [[TwilioStereoTonePlayer audioDevice] pausePlayback];
}

RCT_EXPORT_METHOD(setVolume:(float)volume) {
    NSLog(@"setVolume %f", volume);
    
    _volume = volume;
    
    if (TwilioStereoTonePlayer.audioDevice == nil) {
        NSLog(@"Error: No Custom Audio device found! Are you trying to call setVolume before the Twilio Local Audio Track is created or running in no custom device mode?");
        return;
    }
    
    [[TwilioStereoTonePlayer audioDevice] setPlaybackVolume:volume];
}

RCT_EXPORT_METHOD(setPlaybackSpeed:(float)playbackSpeed) {
    NSLog(@"setPlaybackSpeed %f", playbackSpeed);
    
    _playbackSpeed = playbackSpeed;
    
    if (TwilioStereoTonePlayer.audioDevice == nil) {
        NSLog(@"Error: No Custom Audio device found! Are you trying to call setPlaybackSpeed before the Twilio Local Audio Track is created or running in no custom device mode?");
        return;
    }
    
    [[TwilioStereoTonePlayer audioDevice] setPlaybackSpeed:playbackSpeed];
}

RCT_EXPORT_METHOD(release:(NSString*)filename) {
    NSLog(@"release %@", filename);
    
    if (_loadedFiles) {
        [_loadedFiles removeObjectForKey:filename];
    }
}

RCT_EXPORT_METHOD(terminate) {
    NSLog(@"terminate");
    
    if (_loadedFiles) {
        [_loadedFiles removeAllObjects];

        // We don't need to delete objectives in objective C as ARC will take care of unreferenced objects
        _loadedFiles = NULL;
    }
}


@end

File: android/src/main/java/com/twiliorn/library/TwilioStereoTonePlayer.java
package com.twiliorn.library;

import android.content.res.AssetFileDescriptor;
import android.media.AudioManager;
import android.media.SoundPool;
import android.net.Uri;
import android.util.Log;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

public class TwilioStereoTonePlayer extends ReactContextBaseJavaModule implements SoundPool.OnLoadCompleteListener {

    private static final String TAG = "TwilioStereoTonePlayer";

    // An instance to the SoundPool android class for playing small audio files efficiently
    private SoundPool soundPool;
    private int maxLoadableFiles;

    // A class to store information about a loaded/loading tone file
    private class ToneFile {
        int id = -1;
        boolean loop = false;

        boolean isPlaying = false;

        Promise loadingPromise = null;
    }

    // A map containing FileName and ToneFile instances for all files that are called with preload
    private Map<String, ToneFile> loadedFiles;

    // Properties for currently playing file
    private int currentPlayingFile = -1;
    private float volume = 1.0f;
    private float playbackSpeed = 1.0f;

    ///// Methods
    public TwilioStereoTonePlayer(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return TAG;
    }

    @ReactMethod
    public void initialize(int maxLoadableFiles) {
        this.maxLoadableFiles = maxLoadableFiles;

        loadedFiles = new HashMap<String, ToneFile>(maxLoadableFiles);
        this.soundPool = new SoundPool(
                maxLoadableFiles, AudioManager.STREAM_MUSIC, 0);
        this.soundPool.setOnLoadCompleteListener(this);
    }

    @ReactMethod
    public void preload(String filename, Promise promise) {
        Log.v(TAG,"preload " + filename);

        // Sanity Check
        if (this.soundPool == null || this.loadedFiles == null) {
            promise.reject("Error", "Please call initialize first");
            return;
        }

        if (this.loadedFiles.size() > maxLoadableFiles) {
            promise.reject("Error", "Trying to load more files than you called Initialize with! Already loaded " + maxLoadableFiles + " files.");
        }

        if (this.loadedFiles.containsKey(filename)) {
            promise.resolve(true);
            return;
        }

        // Create an asset file descriptor
        AssetFileDescriptor assetFile;

        // Note: We only try/catch this block to make sure promise.reject is only called once
        // Otherwise its possible for the promise to be rejected in onLoadComplete
        try {
            assetFile = getReactApplicationContext().getAssets().openFd(filename);
        }  catch (Exception e) {
            promise.reject("error", "Unable to load file: " + filename, e);
            return;
        }

        // Generate a ToneFile object
        ToneFile toneFile = new ToneFile();
        toneFile.loadingPromise = promise;

        // Store in the loaded files map
        this.loadedFiles.put(filename, toneFile);

        // SoundPool.Load is an async function and will return its result in onLoadComplete
        toneFile.id = soundPool.load(assetFile, 1);

        // Note: We don't resolve the promises in this function
        // The promises are only resolved in onLoadComplete
    }

    @ReactMethod
    public void play(String filename, boolean isLooping, float volume, float playbackSpeed, Promise promise) {
        Log.v(TAG,String.format("Playing %s isLooping %b volume %f speed %f", filename, isLooping, volume, playbackSpeed));

        // Sanity Check
        if (this.soundPool == null || this.loadedFiles == null) {
            promise.reject("Error", "Please call initialize first");
            return;
        }

        if (!this.loadedFiles.containsKey(filename)) {
            this.preload(filename, promise);

            // If we still don't have the file in the loaded files array, then the promise must have been returned by this.preload so lets just exit
            if (!this.loadedFiles.containsKey(filename)) {
                return;
            }
        }

        ToneFile toneFile = this.loadedFiles.get(filename);

        // We clear the object with the promise when the file is fully loaded, so if the loadingPromise is null that means we are still loading the file
        if (toneFile.loadingPromise != null) {
            promise.reject("Error", "File is not loading yet. " + filename);
            return;
        }

        // Before we play a new tone, lets make sure we pause any tone that is currently playing
        this.pause();

        // Time To Loop. 0 = Loop only once | -1 = Loop infinitely.
        int timesToLoop = 0;
        if (isLooping) {
            timesToLoop = -1;
        }

        this.volume = volume;
        this.playbackSpeed = playbackSpeed;

        this.soundPool.play(toneFile.id, this.volume, this.volume, 1, timesToLoop, this.playbackSpeed);
        currentPlayingFile = toneFile.id;
    }

    @ReactMethod
    public void pause() {
        Log.v(TAG,"Pause");

        // Sanity Check
        if (this.soundPool == null && this.loadedFiles == null)
            return;

        if (currentPlayingFile >= 0) {
            this.soundPool.pause(currentPlayingFile);
            currentPlayingFile = -1;
        }
    }

    @ReactMethod
    public void setVolume(float volume) {
        Log.v(TAG,"setVolume " + volume);

        // Sanity Check
        if (this.soundPool == null && this.loadedFiles == null)
            return;

        this.volume = volume;

        // If we are playing a file when setVolume is called then make sure we update its volume
        if (currentPlayingFile >= 0) {
            this.soundPool.setVolume(currentPlayingFile, this.volume, this.volume);
        }
    }

    @ReactMethod
    public void setPlaybackSpeed(float speed) {
        Log.v(TAG,"setPlaybackSpeed " + speed);

        // Sanity Check
        if (this.soundPool == null && this.loadedFiles == null)
            return;

        if (this.playbackSpeed > 2.0f) {
            Log.w(TAG, "Max Playback Speed of 2.0 is supported, you tried " + this.playbackSpeed);
        } else if (this.playbackSpeed < 0.5f) {
            Log.w(TAG, "Min Playback Speed of 0.5 is supported, you tried " + this.playbackSpeed);
        }

        this.playbackSpeed = speed;

        // If we are playing a file when setVolume is called then make sure we update its volume
        if (currentPlayingFile >= 0) {
            this.soundPool.setRate(currentPlayingFile, this.playbackSpeed);
        }
    }

    @ReactMethod
    public void release(String filename) {
        Log.v(TAG,"release " + filename);

        // Sanity Check
        if (this.soundPool == null && this.loadedFiles == null)
            return;

        if (loadedFiles.containsKey(filename)) {
            ToneFile toneFile = this.loadedFiles.get(filename);

            // If this file was currently playing, then pause the player
            if (currentPlayingFile == toneFile.id) {
                pause();
            }

            // Remove all memory related to this file
            this.soundPool.unload(toneFile.id);
            this.loadedFiles.remove(filename);
        }
    }

    @ReactMethod
    public void terminate() {
        Log.v(TAG,"terminate");

        // Make sure we pause any audio that might be being played right now
        pause();

        // Free all allocated memory
        this.loadedFiles.clear();
        this.loadedFiles = null;

        this.soundPool.release();
        this.soundPool = null;
    }

    //////////// Sound Pool Methods ////////////
    @Override
    public void onLoadComplete(SoundPool soundPool, int id, int status) {

        // Loop through all files in the loaded files map
        Iterator it = loadedFiles.entrySet().iterator();
        while (it.hasNext()) {
            Map.Entry pair = (Map.Entry)it.next();

            // Get the current iterations key (file path), value (ToneFile object)
            String filePath = (String)pair.getKey();
            ToneFile toneFile = (ToneFile)pair.getValue();

            // If its the same file as the one that was just loaded
            if (toneFile.id == id) {
                // Check if the file that was just loaded was loaded successfully
                // Note: Status 0 = success, everything else = failed
                if (status == 0) {
                    // Mark the file to be loaded correctly
                    toneFile.loadingPromise.resolve(true);

                    // Free the reference to the loading promise so garbage collector can clean it later
                    toneFile.loadingPromise = null;
                } else {
                    // Reject the promise
                    toneFile.loadingPromise.reject("Error", "Error loading file with path: " + filePath);
                    toneFile.loadingPromise = null;

                    // Remove the entry of the current file from the loadedFiles map
                    it.remove();
                }

                // As LoadComplete is called once per file, we can just exit the loop
                break;
            }
        }
    }
}